### PR TITLE
Fix issue 20603 - 'cannot use non-constant CTFE pointer in an initializer' in recursive structure with overlap

### PR DESCRIPTION
**Pull request by dkorpel**
The function `hasNonConstPointers` was introduced in https://github.com/dlang/dmd/pull/102 (commit 
https://github.com/dlang/dmd/pull/102/commits/14d37a18e423ef55c05a582c8024990191979b71)

Despite the name, it notably seems to not take const/immutable into account. (Though I'm not sure transitive const already was a thing back in 2011).

This PR does not fix [issue 11268](https://issues.dlang.org/show_bug.cgi?id=11268) because this still fails:
```D
struct A {uint d;}
immutable A abc = { 42 };
immutable(uint)* xyz = &abc.d;
```

However, instead of failing in `initsem.d`, it fails in `todt.d` where `visitAddr(AddrExp e)` only supports a `StructLiteralExp`, not a `DotVarExp` like above.
A `DotVarExp` could be supported in that function, but maybe the `DotVarExp` should be dissected at an earlier stage. I think arrays initializers work this way: all array elements are visited separately instead of the entire array, which is why the 'non-constant CTFE pointer' error was not raised when taking the address of an element.

### CODE REVIEW

**Comment by RazvanN7** [comment]
@dkorpel Are you still pursuing this?

**Comment by dkorpel** [comment]
Yes. The thing that's blocking me is unfamiliarity with DMD's CTFE engine. There's an explicit test checking that "Pointers must not escape from CTFE" (see [test/compilable/interpret3.d](https://github.com/dlang/dmd/pull/10882/files#diff-c9a5d5c93a73af02a901e2d98c5a4083ee8b49450f1352e5b6019c4a9e0ba1b9)) and I'm changing that to only mutable pointers. It looks fine at first glance, but I'm a bit worried I'm overlooking something.

**Comment by RazvanN7** [comment]
I think that the issue is that the CTFE engine rewrites calls to the constructor without understanding that some fields are overlapping and doing so it ends up bypassing the call to the constructor. I looked into [Issue 15938](https://issues.dlang.org/show_bug.cgi?id=15938) which seems to be another manifestation of this fact.

This PR seems to fix the symptom but not the real cause of the issue.

**Comment by dkorpel** [comment]
> This PR seems to fix the symptom but not the real cause of the issue.

Do you still think this PR is an improvement or should it be superseded? It's not just about the union code, the simplest manifestation is the first two lines of the test case:
```D
enum immutable(int)* x = new int(3);
enum const(int)* y = new int(5);
```

Currently gives:
```
onlineapp.d(1): Error: cannot use non-constant CTFE pointer in an initializer `new int(3)`
onlineapp.d(2): Error: cannot use non-constant CTFE pointer in an initializer `new int(5)`
```
While the equivalent array case works.
```D
enum immutable(int)[] a = new int[10];
enum const(int)[] b = new int[10];
```

File: test/compilable/interpret3.d [hunk 1]
```diff
@@ -2473,12 +2473,12 @@ static assert(checkPass("foobar") == 1);
...
 struct Toq
 {
-    const(char)* m;
+    char* m;
 }
 
 Toq ptrRet(bool b)
 {
-    string x = "abc";
+    char[] x = "abc".dup;
```
Comment (RazvanN7) [inline test/compilable/interpret3.d:2481]: Why is this change required?

Comment (RazvanN7) [inline test/compilable/interpret3.d:2481]: Why is this change required?

**Review by RazvanN7 — APPROVED** [review]
This change does make sense on its own. Seeing the union, I instantly thought about the other issue.
